device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java

```diff
@@ -247,12 +252,16 @@ private void stopWorkerThreads()
     {
         if (this.sendTaskScheduler != null)
         {
-            this.sendTaskScheduler.shutdown();
```

Comment:
This was the main culprit. Even though we were calling ```shutdown()``` on the thread pool, the thread pool sometimes stuck around. Switching to ```shutdownNow()``` consistently gives us the termination of all threads in the thread pool like we want.